source/auth_signing_config.c

```diff
@@ -0,0 +1,292 @@
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    if (aws_byte_buf_init(&binding->string_storage, aws_py_get_allocator(), total_string_len)) {
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    binding->native.region.ptr = binding->string_storage.buffer + binding->string_storage.len;
+    binding->native.region.len = region.len;
+    aws_byte_buf_write_from_whole_cursor(&binding->string_storage, region);
+
+    binding->native.service.ptr = binding->string_storage.buffer + binding->string_storage.len;
+    binding->native.service.len = service.len;
+    aws_byte_buf_write_from_whole_cursor(&binding->string_storage, service);
+
+    /* date: store original datetime python object so user doesn't see different timezones after set/get */
+    aws_date_time_init_epoch_secs(&binding->native.date, timestamp);
+    binding->py_date = py_date;
+    Py_INCREF(binding->py_date);
+
+    /* should_sign_param */
+    if (py_should_sign_param_fn == Py_None) {
+        binding->native.should_sign_param = NULL;
+        binding->native.should_sign_param_ud = NULL;
+    } else {
+        binding->native.should_sign_param = s_should_sign_param;
+        binding->native.should_sign_param_ud = binding;
+    }
+    binding->py_should_sign_param_fn = py_should_sign_param_fn;
+    Py_INCREF(binding->py_should_sign_param_fn);
+
+    /* success! */
+    return py_capsule;
+
+error:
+    Py_DECREF(py_capsule);
+    return NULL;
+}
+
+struct aws_signing_config_aws *aws_py_get_signing_config(PyObject *py_signing_config) {
+    struct aws_signing_config_aws *native = NULL;
+
+    PyObject *py_capsule = PyObject_GetAttrString(py_signing_config, "_binding");
+    if (py_capsule) {
+        struct config_binding *binding = PyCapsule_GetPointer(py_capsule, s_capsule_name_signing_config);
+        if (binding) {
+            native = &binding->native;
+            AWS_FATAL_ASSERT(native);
+        }
+        Py_DECREF(py_capsule);
+    }
+
+    return native;
+}
+
+/**
+ * Common start to every getter. Parse arguments and return binding.
+ */
+static struct config_binding *s_common_get(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *py_capsule;
+    if (!PyArg_ParseTuple(args, "O", &py_capsule)) {
+        return NULL;
+    }
+
+    struct config_binding *binding = PyCapsule_GetPointer(py_capsule, s_capsule_name_signing_config);
+    return binding;
+}
+
+PyObject *aws_py_signing_config_get_algorithm(PyObject *self, PyObject *args) {
+    struct config_binding *binding = s_common_get(self, args);
+    if (!binding) {
+        return NULL;
+    }
+
+    return PyLong_FromLong(binding->native.algorithm);
+}
+
+PyObject *aws_py_signing_config_get_credentials_provider(PyObject *self, PyObject *args) {
+    struct config_binding *binding = s_common_get(self, args);
+    if (!binding) {
+        return NULL;
+    }
+
+    Py_INCREF(binding->py_credentials_provider);
+    return binding->py_credentials_provider;
```

Comment:
isn't it naturally going to be incremented when assigned in python land?

Reply:
Returned values need their refcount bumped. Otherwise you'd never be able to return new objects from C to python. Ex: `return PyLong_FromLong(123);`

And it's why you can't just do `return Py_None;`. Instead you use the macro `Py_RETURN_NONE;` which does the refcount bump for you before returning.

Python must decref the returned value once it's in python land. If the user assigns it to a variable that's an INCREF and keeps it alive.